src/dodal/devices/i04/transfocator.py

```diff
@@ -21,38 +31,51 @@ class Transfocator(Device):
     TIMEOUT = 120
     _POLLING_WAIT = 0.01
 
-    def polling_wait_on_stat_rbv(self, for_value):
+    def polling_wait_on_start_rbv(self, for_value):
         # For some reason couldn't get monitors working on START_RBV
         # (See https://github.com/DiamondLightSource/dodal/issues/152)
-        for _ in range(int(self.TIMEOUT / self._POLLING_WAIT)):
+        start_time = time()
+        while time() < start_time + self.TIMEOUT:
             RBV_value = self.start_rbv.get()
-            sleep(self._POLLING_WAIT)
             if RBV_value == for_value:
                 return
-        raise TimeoutError()
+            sleep(self._POLLING_WAIT)
+
+        # last try
+        if self.start_rbv.get() != for_value:
+            raise TimeoutError()
 
-    def set(self, beamsize: float) -> DeviceStatus:
+    def set(self, beamsize_microns: float) -> DeviceStatus:
+        """To set the beamsiz on the transfocator we must:
+        1. Set the beamsize in the calculator part of the transfocator
+        2. Get the predicted number of lenses needed from this calculator
+        3. Enter this back into the device
+        4. Start the device moving
+        5. Wait for the start_rbv goes high and low again
+        """
         subscriber: int
         status = DeviceStatus(self, timeout=self.TIMEOUT)
 
         def set_based_on_predicition(old_value, value, *args, **kwargs):
-            if old_value != value:
-                self.predict_vertical.unsubscribe(subscriber)
+            if not math.isclose(old_value, value, abs_tol=1e-8):
+                self.predicted_vertical_num_lenses.unsubscribe(subscriber)
 
+                # We can only put an integer number of lenses in the beam but the
+                # calculation in the IOC returns the theoretical float number of lenses
                 value = round(value)
                 LOGGER.info(f"Transfocator setting {value} filters")
                 self.number_filters_sp.set(value).wait()
                 self.start.set(1).wait()
-                self.polling_wait_on_stat_rbv(1)
-                self.polling_wait_on_stat_rbv(0)
+                self.polling_wait_on_start_rbv(1)
+                self.polling_wait_on_start_rbv(0)
                 status.set_finished()
```

Comment:
```suggestion
            status.set_finished()
```

Otherwise I think this could deadlock if it gets into this subscriber but doesn't match the `not isclose` condition, since the status would never be set to finished?

Reply:
I don't think we'll end up with a deadlock. The `status` will eventually just timeout in error. It's reasonable to ask if we want that to happen though. I think you're right that actually if the calculation gives the same number of lenses as before we should just immediately set finished. My worry here is that we're assuming the IOC behaves well and doesn't send us a  repeat first, I hope that's a reasonable assumption.

Reply:
Ah ok and the timeout doesn't cause an error? Fair enough, I assumed that if it timed out that would cause the scan to fail (or if there wasn't a timeout that it would deadlock).

Reply:
Sorry, you're right it will cause an error. In which case this is required